File: macos-app/Lexume/Reader/PlayerBarView.swift
import SwiftUI

/// Floating playback controls: shown below the reader when audio is
/// available (or being generated), hidden entirely when audio mode is "off".
@MainActor
struct PlayerBarView: View {
    let vm: ReaderViewModel

    var body: some View {
        VStack(spacing: 8) {
            if let audioError = vm.audioError {
                Text(audioError)
                    .font(.caption)
                    .foregroundStyle(.red)
                    .lineLimit(2)
            }

            if vm.hasAudio {
                playbackControls
            } else if vm.isGeneratingAudio {
                HStack(spacing: 8) {
                    ProgressView().controlSize(.small)
                    Text("Generating narration…")
                        .font(.callout)
                        .foregroundStyle(.secondary)
                }
            } else {
                Button {
                    vm.requestGenerateAudio()
                } label: {
                    Label("Generate Audio", systemImage: "waveform")
                }
                .controlSize(.large)
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 10)
        .frame(maxWidth: .infinity)
        .background(.regularMaterial)
    }

    private var playbackControls: some View {
        VStack(spacing: 6) {
            progressBar
            HStack(spacing: 16) {
                Text(formattedTime(vm.playbackEngine.currentTime))
                    .font(.caption.monospacedDigit())
                    .foregroundStyle(.secondary)
                    .frame(width: 40, alignment: .leading)

                Spacer()

                Button { vm.playbackEngine.restart() } label: {
                    Image(systemName: "arrow.counterclockwise")
                }
                .buttonStyle(.plain)

                Button { vm.playbackEngine.skip(by: -15) } label: {
                    Image(systemName: "gobackward.15")
                        .font(.system(size: 16))
                }
                .buttonStyle(.plain)

                Button {
                    vm.playbackEngine.toggle()
                } label: {
                    Image(systemName: vm.playbackEngine.isPlaying ? "pause.circle.fill" : "play.circle.fill")
                        .font(.system(size: 32))
                }
                .buttonStyle(.plain)

                Button { vm.playbackEngine.skip(by: 15) } label: {
                    Image(systemName: "goforward.15")
                        .font(.system(size: 16))
                }
                .buttonStyle(.plain)

                Spacer()

                Text(formattedTime(vm.playbackEngine.duration))
                    .font(.caption.monospacedDigit())
                    .foregroundStyle(.secondary)
                    .frame(width: 40, alignment: .trailing)
            }
        }
    }

    private var progressBar: some View {
        GeometryReader { geo in
            let progress = vm.playbackEngine.duration > 0 ? vm.playbackEngine.currentTime / vm.playbackEngine.duration : 0
            ZStack(alignment: .leading) {
                Capsule().fill(Color.secondary.opacity(0.25))
                Capsule().fill(Color.accentColor)
                    .frame(width: max(0, geo.size.width * progress))
            }
            .frame(height: 6)
            .contentShape(Rectangle())
            .gesture(
                DragGesture(minimumDistance: 0)
                    .onChanged { value in seek(to: value.location.x, in: geo.size.width) }
                    .onEnded { value in seek(to: value.location.x, in: geo.size.width) }
            )
        }
        .frame(height: 16)
    }

    private func seek(to x: CGFloat, in width: CGFloat) {
        guard vm.playbackEngine.duration > 0, width > 0 else { return }
        let fraction = max(0, min(1, x / width))
        vm.playbackEngine.seek(to: fraction * vm.playbackEngine.duration)
    }

    private func formattedTime(_ seconds: Double) -> String {
        guard seconds.isFinite, seconds >= 0 else { return "0:00" }
        let total = Int(seconds)
        return String(format: "%d:%02d", total / 60, total % 60)
    }
}

File: android-app/app/src/main/java/com/aboooooo57/lexume/ui/reader/PlayerBarView.kt
package com.aboooooo57.lexume.ui.reader

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.FastForward
import androidx.compose.material.icons.filled.FastRewind
import androidx.compose.material.icons.filled.GraphicEq
import androidx.compose.material.icons.filled.Pause
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Replay
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp

/**
 * Floating playback controls - shown below the reader when audio is
 * available (or being generated), hidden entirely when audio mode is "off".
 * Mirrors `Reader/PlayerBarView.swift`, with a Material3 [Slider] standing
 * in for the Swift version's hand-rolled `GeometryReader`+`DragGesture`
 * progress bar - Slider already gives drag-to-seek for free, no need to
 * reimplement that gesture logic here.
 */
@Composable
fun PlayerBarView(
    hasAudio: Boolean,
    isGeneratingAudio: Boolean,
    audioError: String?,
    playbackEngine: PlaybackEngine,
    onGenerateAudio: () -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 10.dp)
    ) {
        if (audioError != null) {
            Text(
                audioError,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.error,
                maxLines = 2,
                modifier = Modifier.padding(bottom = 6.dp)
            )
        }

        when {
            hasAudio -> PlaybackControls(playbackEngine)
            isGeneratingAudio -> LoadingRow("Generating narration…")
            else -> Button(onClick = onGenerateAudio) {
                Icon(Icons.Filled.GraphicEq, contentDescription = null)
                Spacer(Modifier.width(8.dp))
                Text("Generate Audio")
            }
        }
    }
}

@Composable
private fun LoadingRow(message: String) {
    Row(verticalAlignment = Alignment.CenterVertically) {
        CircularProgressIndicator(
            modifier = Modifier
                .height(16.dp)
                .width(16.dp),
            strokeWidth = 2.dp
        )
        Spacer(Modifier.width(8.dp))
        Text(message, style = MaterialTheme.typography.bodyMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
    }
}

@Composable
private fun PlaybackControls(engine: PlaybackEngine) {
    Column {
        Slider(
            value = engine.currentTime.toFloat(),
            onValueChange = { engine.seek(it.toDouble()) },
            valueRange = 0f..maxOf(engine.duration.toFloat(), 0.01f)
        )
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                formatTime(engine.currentTime),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.width(40.dp)
            )
            Spacer(Modifier.weight(1f))
            IconButton(onClick = { engine.restart() }) {
                Icon(Icons.Filled.Replay, contentDescription = "Restart")
            }
            IconButton(onClick = { engine.skip(-15.0) }) {
                Icon(Icons.Filled.FastRewind, contentDescription = "Back 15 seconds")
            }
            IconButton(onClick = { engine.toggle() }) {
                Icon(
                    if (engine.isPlaying) Icons.Filled.Pause else Icons.Filled.PlayArrow,
                    contentDescription = if (engine.isPlaying) "Pause" else "Play",
                    modifier = Modifier
                        .height(32.dp)
                        .width(32.dp)
                )
            }
            IconButton(onClick = { engine.skip(15.0) }) {
                Icon(Icons.Filled.FastForward, contentDescription = "Forward 15 seconds")
            }
            Spacer(Modifier.weight(1f))
            Text(
                formatTime(engine.duration),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.width(40.dp)
            )
        }
    }
}

private fun formatTime(seconds: Double): String {
    if (!seconds.isFinite() || seconds < 0) return "0:00"
    val total = seconds.toInt()
    return "%d:%02d".format(total / 60, total % 60)
}
